document.go

```diff
@@ -104,6 +104,29 @@ func (doc *Document) SetAll(values map[string]interface{}) {
 	}
 }
 
+// GetAll returns a map of all available fields in the document. Nested fields are represented by sub-maps. This is a deep copy, but values are note cloned.
+func (doc *Document) GetAll() map[string]interface{} {
+	return copyMap(doc.fields)
+}
+
+// GetKeys returns a slice of all available field names in the document. Nested fields are represented using dot notation (in such case parent fields are not represented standalone). There is no guarantee on the order.
+func (doc *Document) GetKeys() []string {
+	return getKeysRecursive(doc.fields, "")
+}
+
+func getKeysRecursive(fields map[string]interface{}, prefix string) []string {
```

Comment:
I think this can be implemented without the need of the extra `prefix` parameter, by concatenating `key + "."` to each element of `result`

Reply:
I think communicating this information to sub-calls is mandatory in such a recursive implementation. I might be wrong, but I'm afraid this is "mathematically" unavoidable.  Otherwise level n call has no way to know about the parent nodes that have been visited during level 0..n-1 calls.

That being said, I made it basic and recursive for the sake of simplicity, and because the depth of the traversal will always be marginal which doesn't justify the overhaul of an iterative version. But of course we could make it iterative and avoid an extra parameter.

BTW I moved it to utils.go, although it could be made more generic (specify separator, avoid having to pass the empty prefix at first call, etc.)

Reply:
Hi Stefano,

It works, but this is pretty ineffective to re-scan each sub-map. Transferring the prefix just takes a bit on the stack but has no additional computational time. That being said, that's your project, so you're the boss :) But to me performance is key in such a lib.

As for the name, you already have a getKeys() function in compare.go with the same signature. This one doesn't deep-traverse the map, and sorts the result. I'm not fond of the "Recursive" prefix as well but...

Reply:
If we were to be so performance oriented, I think we would't go for a recursive implementation at all, so it is a bit overkilled to me thinking at optimization at this level. Moreover, we are not talking of a function affecting query execution time, so it's fine to trade some cpu cycle for simplicity.

Regarding the function name, you can call it `getMapKeys()`